File: anaconda.py
import sys
print sys.argv[1]
source = file(sys.argv[1])
lines = source.readlines()
source.close()

# x = 5
# y = x * x
# print y

def interpret(lines):
    assignment_hash = {}
    output = []
    for line in lines:
      if "=" in line:
          lhs, rhs = line.split("=")
          assignment_hash[lhs.strip()] = evaluate(rhs.strip(), assignment_hash)
      elif "print" in line:
          print_statement = evaluate(line[6:], assignment_hash)
          output.append(print_statement)
    print output

def is_int(expression):
    try: int(expression)
    except ValueError:
       return False
    return True

def evaluate(expression, assignment_hash):
  expression = expression.strip()
  if is_int(expression):
      return int(expression)
  elif expression in assignment_hash:
      return assignment_hash[expression]
  else:
      operand1, operator, operand2 = expression.partition('*')
      return evaluate(operand1, assignment_hash) * evaluate(operand2, assignment_hash)

interpret(lines)

File: anaconda.rb
# Script to interpret a basic theoretical language ("anaconda") that uses only simple variable assignments, binary operators and print statements
# When run with program.an, should return 25

def interpret(lines)
  assignment_hash = {}
  output = []
  lines.each do |line|
    if line.include? "="
      lhs, rhs = line.split("=").map {|char| char.strip}
      assignment_hash[lhs] = evaluate(rhs, assignment_hash)
    elsif line.include? "print"
      m = line.match(/print /)
      output << evaluate(m.post_match, assignment_hash)
    end
  end
  output
end

def evaluate(expression, assignment_hash)
  expression.strip!
  if expression =~ /^\d+$/
    expression.to_i
  elsif expression =~ /^\w+$/
    raise Error unless assignment_hash.has_key?(expression)
      assignment_hash[expression] 
  elsif matchdata = expression.match(/[-+*\/]/)
    operator = matchdata.to_s 
    operands = expression.split(operator)
    operand1, operand2 = operands.map {|operand| evaluate(operand, assignment_hash)}
    case operator
      when "+"
        operand1 + operand2
      when "-"
        operand1 - operand2
      when "*"
        operand1 * operand2
      when "/"
        operand1 / operand2
    end
  end
end

ARGV.each do |filename|
  lines = File.readlines(filename)
  puts interpret(lines).join("\n")
end
